### PR TITLE
AWS SAM package improvements

### PR DESCRIPTION
**Pull request by munichbughunter**
* Connect to tcp socket if you wanna build the sam image in a GitLab CI environment
* Improve code documentation
* Add README.md to give an overview how to use the package locally and also in a GitLab environment

Closes https://github.com/dagger/dagger/issues/2449

### CODE REVIEW

**Review by gerhard — CHANGES_REQUESTED** [review]
This is a great PR with an amazing README, thank you @munichbughunter!

Apart from a few small changes for which I have left inline comments, there is a more significant change which would be greatly appreciated: could we move the content of the README.md into a doc page? I think it would go nicely under the **Use Cases** section, but **References** also works. WDYT @jpadams?

The Lint & DCO checks that are failing are self-explanatory. FWIW: https://docs.dagger.io/1227/contributing/#how-to-run-linters-locally

One important thing worth mentioning is that keeping `cue` snippets in separate files makes working with them easier. This is a good & recent example: https://raw.githubusercontent.com/dagger/dagger/v0.2.28/docs/core-concepts/1247-dagger-fs.md . I think that this will be helpful for those links: https://docs.dagger.io/1227/contributing#docs . This is also relevant: https://github.com/dagger/dagger/blob/main/docs/README.md .

Hope the above helps! This is not far off from getting merged.

File: pkg/universe.dagger.io/alpha/aws/sam/README.md [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+#### Plan for a .zip archive
+
+This is a the plan for a `.zip archives` function. 
+
+```cue title="samZip.cue"
+package myAwesomeSamPackage
+
+import (
+    "dagger.io/dagger
+    "universe.dagger.io/x/munichbughunter@gmail.com/sam
```
[the remark this answers is not in the document]
Comment (munichbughunter) [reply pkg/universe.dagger.io/alpha/aws/sam/README.md:46]: You are right. Fixed.

File: pkg/universe.dagger.io/alpha/aws/sam/README.md [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+
+This is a the plan for a `docker image` function. 
+In case of building a docker image we have to define the docker socket and we don't need the S3 bucket anymore. 
+
+```cue title="samImage.cue"
+package myAwesomeSamPackage
+
+import (
+    "dagger.io/dagger
+    "universe.dagger.io/x/munichbughunter@gmail.com/sam
```
[the remark this answers is not in the document]
Comment (munichbughunter) [reply pkg/universe.dagger.io/alpha/aws/sam/README.md:100]: You are right. Fixed.

File: pkg/universe.dagger.io/alpha/aws/sam/README.md [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+If you plan to run the plan with the docker image in a GitLab CI environment you have to make small changes to get everything up and running. This comes because on GitLab you have to use a `DinD-Service` and you can not connect via `docker socket` with it you have to use the `tcp-socket`.
+
+First we have to change the plan itself to use `tcp-socket` if it is executed in the GitLab environment. 
+
+```cue title="samImage.cue"
+package myAwesomeSamPackage
+
+import (
+    "dagger.io/dagger
+    "universe.dagger.io/x/munichbughunter@gmail.com/sam
```
[the remark this answers is not in the document]
Comment (munichbughunter) [reply pkg/universe.dagger.io/alpha/aws/sam/README.md:221]: Fixed.

File: pkg/universe.dagger.io/alpha/aws/sam/sam.cue [hunk 4]
```diff
@@ -5,36 +5,60 @@ import (
 	"universe.dagger.io/docker"
 )
 
+// Working directory in Sam image
 _destination: "/var/task"
 
+// AWS Sam image
 #DefaultImage: docker.#Pull & {
 	source: "amazon/aws-sam-cli-build-image-provided:latest"
 }
```
[the remark this answers is not in the document]
Comment (munichbughunter) [reply pkg/universe.dagger.io/alpha/aws/sam/sam.cue:14]: I get your point. When I take a look on the official
[Amazon Web Services Sam build image](https://hub.docker.com/r/amazon/aws-sam-cli-build-image-provided/tags) 

I only find one tag called latest.

File: pkg/universe.dagger.io/alpha/aws/sam/README.md [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+
+:bulb: Don't forget to set the needed environment variables in your GitLab CI environment. 
+
+## :handshake: Contributing
+
+If you have a specific need, don't hesitate to write an [issue](https://github.com/munichbughunter/dagger) or create a [pull request]((https://github.com/munichbughunter/dagger/pulls)) on our repository! :rocket:
+
+See the workflow below to contribute.
+
+> :bulb: Check that [post](https://chris.beams.io/posts/git-commit/) to learn how write good commit message
```
[the remark this answers is not in the document]
Comment (munichbughunter) [reply pkg/universe.dagger.io/alpha/aws/sam/README.md]: I linked to the official contribution guide.

File: pkg/universe.dagger.io/alpha/aws/sam/sam.cue (hunk 4 above)
[the remark this answers is not in the document]
Comment (gerhard) [reply pkg/universe.dagger.io/alpha/aws/sam/sam.cue:14]: In that case, I suggest that we link to the SHA256 instead. I can make the change, no worries.

File: pkg/universe.dagger.io/alpha/aws/sam/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (gerhard) [reply pkg/universe.dagger.io/alpha/aws/sam/README.md]: That is a step in the right direction, for sure.

Let me move this README to docs and make the last change. I will wait for your sign-off before merging this.

File: pkg/universe.dagger.io/alpha/aws/sam/sam.cue (hunk 4 above)
[the remark this answers is not in the document]
Comment (gerhard) [reply pkg/universe.dagger.io/alpha/aws/sam/sam.cue]: Resolved in https://github.com/dagger/dagger/pull/2890/commits/cc235cd05239ca8a2c295835c9d00a44e952cb52

**Comment by gerhard** [comment]
@munichbughunter can you please double-check that everything looks right?

If it does to you, I can do another review and go for merge.

Almost there!

**Review by gerhard — APPROVED** [review]
These changes are ready to be merged from my perspective. We are just waiting on @munichbughunter to confirm that my last commits stayed true to the initial intent of this PR. When that happens, this is ready to **Squash and merge** (all checks should be green before we do that). If this doesn't happen today, I'm leaving it with you @grouville. Thanks!

**Review by munichbughunter — COMMENTED** [review]
@gerhard looks good for me - from my side it can be merged. 
Thank you for the support.

**Pull request merged**